### DISCUSSION TITLE
Completion of the Sync api

### DISCUSSION THREAD

**Discussion by CuriousCrook** [FreeOpcUa/opcua-asyncio#682] [Q&A]
Hello all,

I use the "old" python client python-opcua to test certain processes on a server with a somewhat complex information model. In my case, the flows have always been purely sequential so far. For my scripts, using the asyncio client leads to more effort, more code, more confusing programs.

However, it looks like the old toolkit is no longer maintained to the same extent as the asyncio toolkit. For me the "sync api" in the asyncio client is a great way to combine the advantages of both toolkits. 

Unfortunately the API is not "ready enough" for many use cases. I just tried to connect to a server, which
- authenticates users via certificates
- Checks the application uri against the one in the application certificate

Both was not possible at first, because with in the sync api the methods
- client.load_private_key()
- client.load_client_certificate()
 
and the property 
- client.application_uri
 
were missing.

I think the sync api is a good chance to move more developers from the old toolkit to the new one. So my question, what are the further plans for the sync api? Will it ever be completed?

With kind regards
Curious Crook

**Comment by oroulet** [comment]
just add them to the asyncua/sync.py and make a PR. you will see it is crazy easy to add functoins there. just copy the declaration and write `pass`

**Reply by CuriousCrook** [reply]
That sounds like a plan. I will do it.
But there is one place where I will need help.

I unsuccessfully tried to add UaFile (./client/ua_file.py) to the sync api. Maybe someone can help me there?

**Reply by oroulet** [reply]
Ua_file is a bit more complicated. Start by making a class SyncUaFile taking an UaFile as argument and defining all methods. Then we need to modify the decorators to also support them

**Reply by CuriousCrook** [reply]
Okay, I'm going to try that. But first I want to complete "Part 5: Information Model - Annex C (normative) File Transfer". 

The existing file transfer in ./client/ua_file.py can only read small files. The functionality of FileDirectoryType is currently not considered either. I wish to have a ua-like interface and a convenient one to work similar to local files.

**Reply by oroulet** [reply]
You are welcome. I have never had a need for these things

**Reply by CuriousCrook** [reply]
File Transfer has been ready for months. I use it for file transfer with a Softing OPC UA server. 

See "Add low-level file system functionality #696".

Unfortunately I can't manage to include UaFile and UaDirectory in the sync interface.

I already looked at the classes "Client" and "SyncNode" and tried to implement UaFile similarly. Since I have no idea what is being done in sync.py, I am not getting anywhere.

Can anyone help me?

If I could get an example for UaFile, I could probably implement UaDiretory myself.

**Comment by CuriousCrook** [comment]
Can someone help me to include UaFile and UaDirectory in the Sync API. I have looked at Sync.py. Unfortunately I have no idea what is happening. (No experience with asyncio)
